File: code/剑指offer51.py
class Solution:
    def reversePairs(self, nums) -> int:
        def mergeSort(tmp,start,end):
            if start>=end:
                return 0
            mid = start+(end-start)//2
            count = mergeSort(tmp,start,mid)+mergeSort(tmp,mid+1,end)
            i,j,pos = start,mid+1,start
            while i<=mid and j<=end:
                if nums[i]<=nums[j]:
                    tmp[pos] = nums[i]
                    i+=1
                    count += (j-(mid+1))
                else:
                    tmp[pos] = nums[j]
                    j+=1
                pos+=1
            while i<=mid:
                tmp[pos] = nums[i]
                count+=(j-(mid+1))
                i+=1
                pos+=1
            while j<=end:
                tmp[pos] = nums[j]
                pos+=1
                j+=1
            nums[start:end+1] = tmp[start:end+1]
            return count
        tmp = [0]*len(nums)
        res = mergeSort(tmp,0,len(nums)-1)
        return res

File: code/java/剑指offer51.java
class Solution {

    public int mergeSort(int[] tmp,int[] nums,int left,int right){
        if(left>=right) return 0;

        int mid = (left+right)/2;
        int res = mergeSort(tmp,nums,left,mid)+mergeSort(tmp,nums,mid+1,right);

        int i = left;
        int j = mid+1;
        int index = left;
        while(i<=mid&&j<=right){
            if(nums[i]<=nums[j]){
                tmp[index++] = nums[i++];
                res+=(j-(mid+1));
            }
            else tmp[index++] = nums[j++];
        }
        while(i<=mid){
            tmp[index++] = nums[i++];
            res+=(j-(mid+1));
        }
        while(j<=right) tmp[index++] = nums[j++];
        for(int k=left;k<=right;k++) nums[k] = tmp[k];
        return res;
    }

    public int reversePairs(int[] nums) {
        int[] tmp = new int[nums.length];
        int res = mergeSort(tmp,nums,0,nums.length-1);
        return res;
    }
}